op-node: Remove Unsafe Sync Target & Direct Engine Manipulation

trianglesphere opened the pull request:
**Description**

This PR primarily moves the last of the direct execution engine manipulation from the EngineQueue to the EngineController. It also removes the unsafe sync target because maintaining the unsafe sync target was difficult and will be replaced by a different EL sync method.


**Tests**

Again no tests because this is tidying.

ajsutton commented on op-e2e/actions/sync_test.go:
```diff
@@ -158,8 +158,8 @@ func TestUnsafeSync(gt *testing.T) {
 		// Check engine sync target updated.
-		require.Equal(t, sequencer.L2Unsafe().Hash, sequencer.EngineSyncTarget().Hash)
-		require.Equal(t, verifier.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
+		require.Equal(t, sequencer.L2Unsafe().Hash, sequencer.L2Unsafe().Hash)
+		require.Equal(t, verifier.L2Unsafe().Hash, verifier.L2Unsafe().Hash)
```
I'm not sure it makes sense to assert that the unsafe head matches the unsafe head.

trianglesphere replied:
I'll actually look at this test - it was trying to test this part of snap sync so maybe it's not needed anymore

trianglesphere replied:
Ok, I updated the tests & left comments on what they are supposed to be doing.

ajsutton commented on op-e2e/actions/sync_test.go:
```diff
@@ -197,9 +197,9 @@ func TestEngineP2PSync(gt *testing.T) {
+		require.NotEqual(t, verifier.L2Unsafe().Hash, verifier.L2Unsafe().Hash)
 		require.Equal(t, verifier.L2Unsafe().Hash, verifierUnsafeHead.Hash)
-		require.Equal(t, sequencer.L2Unsafe().Hash, verifier.EngineSyncTarget().Hash)
+		require.Equal(t, sequencer.L2Unsafe().Hash, verifier.L2Unsafe().Hash)
```
Also seems like a tautology.

ajsutton commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -248,18 +241,16 @@ func (eq *EngineQueue) PendingSafeL2Head() eth.L2BlockRef {
-		return eq.tryUpdateEngine(ctx)
+	// If we don't need to call FCU, keep going b/c this was a no-op. If we needed to
+	// perform a network call, then we should yield even if we did not encounter an error.
+	if err := eq.ec.TryUpdateEngine(ctx); err != errNoFCUNeeded {
```
```suggestion
	if err := eq.ec.TryUpdateEngine(ctx); errors.Is(err, errNoFCUNeeded) {
```

I may have the args the wrong way around of `errors.Is` but it gives us the flexibility to wrap the returned error to provide more detail if needed in the future.

trianglesphere replied:
good callout

ajsutton commented on op-node/rollup/derive/engine_queue.go:
```diff
@@ -616,12 +522,11 @@ func (eq *EngineQueue) consolidateNextSafeAttributes(ctx context.Context) error
 	// unsafe head stays the same, we did not reorg the chain.
 	eq.safeAttributes = nil
 	eq.logSyncProgress("reconciled with L1")
+	eq.postProcessSafeL2()
```
Seems suspicious that this moved out of the `isLastInSpan` if block.

trianglesphere replied:
You're right - it pulls data from two different sources (the L2 safe head & the origin of the pipeline) so it's not safe to call in the middle of a span batch update.

trianglesphere replied:
gated behind a check. Also FYI we know call `postProcessSafeL2` when forcing safe attributes.

ajsutton commented on op-service/eth/sync_status.go:
```diff
@@ -37,7 +37,4 @@ type SyncStatus struct {
 	// UnsafeL2SyncTarget points to the first unprocessed unsafe L2 block.
 	// It may be zeroed if there is no targeted block.
 	UnsafeL2SyncTarget L2BlockRef `json:"queued_unsafe_l2"`
-	// EngineSyncTarget points to the L2 block that the execution engine is syncing to.
-	// If it is ahead from UnsafeL2, the engine is in progress of P2P sync.
-	EngineSyncTarget L2BlockRef `json:"engine_sync_target"`
```
Strictly this is a breaking API change. It's probably not a big deal but just wanted to confirm we're ok with that.

trianglesphere replied:
Yea, it's intentional. This part of the code was always going to disappear as we're going to unify the unsafe head & the engine sync target.

trianglesphere committed:
op-node: Remove Unsafe Sync Target & Direct Engine Manipulation

This fully moves all execution engine manipulation into the Engine Controller
from the Engine Queue. It does not remove the proxy methods however.

pcw109550 commented on op-node/rollup/derive/engine_controller.go:
```diff
@@ -165,7 +164,6 @@ func (e *EngineController) ConfirmPayload(ctx context.Context) (out *eth.Executi
-	e.syncTarget = ref
 
 	e.metrics.RecordL2Ref("l2_unsafe", ref)
 	e.metrics.RecordL2Ref("l2_engineSyncTarget", ref)
```
Need to remove this as well.
